### PR TITLE
OSX codesign binary fix and par.pl fix for codesigned OSX binaries

### PR DESCRIPTION
**Pull request by plk**
This pull request adds a small exe for OSX which edits in-place a `pp` created exe to correct the Mach-O headers so that the OSX `codesign` utility will work.

There is also a small patch to `par.pl` so that `Archive::Zip` can find the EOCD in codesigned binaries. `codesign` adds around 180K to the binary and with the default 32K ChunkSize for `Archive::Zip` is not enough to allow the algorithm to find the EOCD in codesigned binaries. With this small fix, it works fine.

This will become a major issue soon as Apple is planning to make codesigning mandatory.

### CODE REVIEW

File: contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/* -*- C -*- main.c */
+
```
Comment (rschupp) [inline contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c]: Please add a simple copyright and license note, e.g. the first 4 paragraphs of the LICENSE file
would do (of course at your disgression, but don't forget to change the copyright).

File: contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        printf("  New VM Size: %i\n", (int)segment->vmsize);
+        write_bytes(obj_file, current_offset, sizeof(struct segment_command_64), segment);
+        free(segment);
+      }
+    }
+
+    /* 
+       LC_SYMTAB.String Table Size = .exe size - String Table Offset
+    */ 
+    if (cmd->cmd == LC_SYMTAB) {
```
Comment (rschupp) [inline contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c]: Nit: can you change "if" to "else if" here. This makes it obvious the following block and the previous one
are mutually exclusive.

File: contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      printf("Correcting LC_SYMTAB\n");
+      printf("  Old String Table Size: %i\n", (int)symtab->strsize);
+      symtab->strsize = exesize - symtab->stroff;
+      printf("  New String Table Size: %i\n", (int)symtab->strsize);
+      write_bytes(obj_file, current_offset, sizeof(struct symtab_command), symtab);
+      free(symtab);
+    }
+    
+    current_offset += cmd->cmdsize;
+  }
```
Comment (rschupp) [inline contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c]: Nit: add "free(cmd);

File: contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c (hunk 1 above)
Comment (rschupp) [inline contrib/pp_osx_codesign_fix/pp_osx_codesign_fix.c]: Please add a comment containing the command to build the pp_osx_codesign_fix executable.

**Comment by plk** [comment]
Tidied up a bit, should be ok now.

**Comment by plk** [comment]
@rschupp - can we bump the signature window up again and do another release? I have just build a new project that required more `--link` statements and this increased the codesign appended data to 271K and it's broken again ... I think it's the checksum data that increases with the file size so how about bumping the `Archive::Zip` ChunkSize to 1024K? Or perhaps even making it configurable ... I don't think the `PAR.pm` signature matters as that seems to be an exe _creation_ time setting (and there is also a creation time setting for the ChunkSize in `par.pl` which does not help the extraction issue).

**Comment by plk** [comment]
@rschupp - I have opened a new pull request - this fix turns out not to be general enough when the binary size increases - see new pull request - this should work for any size.

**Pull request merged**